[Feature][runtime] Support the use of Python ChatModel in Java

GreatEugenius opened the pull request:
<!--
* Thank you very much for contributing to Flink Agents.
* Please add the relevant components in the PR title. E.g., [api], [runtime], [java], [python], [hotfix], etc.
-->

<!-- Please link the PR to the relevant issue(s). Hotfix doesn't need this. -->
Linked issue: #313

### Purpose of change
Support the use of Python ChatModel in Java Flink Agents jobs.
<!-- What is the purpose of this change? -->

### Tests
ut
<!-- How is this change verified? -->

### API
no
<!-- Does this change touches any public APIs? -->

### Documentation

<!-- Do not remove this section. Check the proper box only. -->

- [x] `doc-needed` <!-- Your PR changes impact docs -->
- [ ] `doc-not-needed` <!-- Your PR changes do not impact docs -->

GreatEugenius commented:
@Sxnan @wenjin272 please help take a look this PR, thanks.

wenjin272 commented on plan/src/main/java/org/apache/flink/agents/plan/resourceprovider/PythonResourceProvider.java:
```diff
@@ -46,6 +54,25 @@ public PythonResourceProvider(
         this.module = module;
         this.clazz = clazz;
         this.kwargs = kwargs;
+        this.descriptor = null;
```
Should we add `descriptor` field for  `PythonResourceProvider` deserialized from agent plan declared using python api?

wenjin272 commented on plan/src/main/java/org/apache/flink/agents/plan/serializer/ResourceProviderJsonDeserializer.java:
```diff
@@ -76,6 +76,15 @@ public ResourceProvider deserialize(
+        try {
+            ResourceDescriptor descriptor =
+                    mapper.treeToValue(node.get("descriptor"), ResourceDescriptor.class);
+            if (descriptor != null) {
```
We can check the `descriptor` field existence rather than if the field is null.

wenjin272 reviewed (commented):
Hi, @GreatEugenius, thanks for your work. Overall looks good to me . I left some comments about details.

Besides, I think this pr should add some test cases to show how to declare an agent in java using python chat model, and verify the behavior.

Sxnan commented on api/src/main/java/org/apache/flink/agents/api/prompt/Prompt.java:
```diff
@@ -76,6 +77,14 @@ public String formatString(Map<String, String> kwargs) {
                 });
     }
 
+    public List<ChatMessage> formatMessages(String defaultRoleValue, Map<String, Object> kwargs) {
```
Why do we need to pass the Object type as the value here? Can't we reuse the `formatMessages(MessageRole defaultRole, Map<String, String> kwargs)`?

This method is not used in Java. We should have documentation to say this is intended to be used by Python code.

Sxnan commented on api/src/main/java/org/apache/flink/agents/api/resource/python/PythonChatModelConnection.java:
```diff
@@ -0,0 +1,63 @@
+    private PyObject chatModel;
+
+    public PythonChatModelConnection(
+            PythonResourceAdapter adapter,
```
The adapter is not used

Sxnan commented on api/src/main/java/org/apache/flink/agents/api/resource/python/PythonChatModelSetup.java:
```diff
@@ -0,0 +1,76 @@
+    private final PyObject chatModelSetup;
+    private final PythonResourceAdapter adapter;
+
+    public PythonChatModelSetup(
```
Where is this class used?

Sxnan commented on api/src/main/java/org/apache/flink/agents/api/resource/python/PythonResourceAdapter.java:
```diff
@@ -0,0 +1,77 @@
+     * @param message the Java ChatMessage to convert
+     * @return the Python representation of the chat message
+     */
+    Object toPythonChatMessage(ChatMessage message);
```
Does `ChatMessage` only make sense to ChatModel? If so, we should not put the method in this General Resource Adapter interface

Sxnan commented on plan/src/main/java/org/apache/flink/agents/plan/AgentPlan.java:
```diff
@@ -76,6 +79,8 @@ public class AgentPlan implements Serializable {
 
     private AgentConfiguration config;
 
+    private PythonResourceAdapter pythonResourceAdapter;
```
Is this serializable? Should it be transient?

Sxnan reviewed (commented):
Thanks for the PR. However, the PR lacks tests. For such a complex feature, we need to have unit tests and e2e tests.

GreatEugenius commented:
Hi @Sxnan and @wenjin272, thank you for the review! I have updated this PR and added a PythonChatModel example for the end-to-end test. Please take a look. Thanks!

Sxnan commented on e2e-test/integration-test/src/main/java/org/apache/flink/agents/integration/test/AgentWithPythonChatModel.java:
```diff
@@ -0,0 +1,144 @@
+        return ResourceDescriptor.Builder.newBuilder(PythonChatModelSetup.class.getName())
+                .addInitialArgument("connection", "pythonChatModelConnection")
+                .addInitialArgument(
+                        "module", "flink_agents.integrations.chat_models.tongyi_chat_model")
```
Should we use ollama model for e2e test?

GreatEugenius commented:
Hi @Sxnan, thank you for the review! I have updated this PR; please review it again.

xintongsong commented on api/src/main/java/org/apache/flink/agents/api/chat/model/python/PythonChatModelConnection.java:
```diff
@@ -0,0 +1,74 @@
+    @Override
+    public ChatMessage chat(
+            List<ChatMessage> messages, List<Tool> tools, Map<String, Object> arguments) {
+        throw new UnsupportedOperationException(
```
Why is this not supported?

xintongsong commented on plan/src/main/java/org/apache/flink/agents/plan/resourceprovider/PythonResourceProvider.java:
```diff
@@ -63,11 +90,21 @@ public Map<String, Object> getKwargs() {
+                    constructor.newInstance(
+                            pythonResourceAdapter, pyResource, descriptor, getResource);
+        }
+        throw new UnsupportedOperationException("PythonResourceAdapter is not set.");
```
Use `checkState`

GreatEugenius committed:
[Feature][runtime] Support the use of Python ChatModel in Java

GreatEugenius committed:
Remove unnecessary dependency

xintongsong reviewed (approved):
LGTM
